File: Week_07/G20200343040009/Java/Trie.java
/**
 * LeetCode208
 * 实现前缀树
 */
public class Trie {
    private class TreeNode {
        private boolean isEnd;
        private TreeNode[] next;

        public TreeNode() {
            isEnd = false;
            next = new TreeNode[26];
        }
    }

    private TreeNode root;

    public Trie() {
        this.root = new TreeNode();
    }

    public void insert(String word) {
        TreeNode node = root;
        for (int i = 0; i < word.length(); i++) {
            int currChar = word.charAt(i) - 'a';
            if (node.next[currChar] == null) {
                node.next[currChar] = new TreeNode();
            }
            node = node.next[currChar];
        }
        // 标记为是一个单词
        node.isEnd = true;
    }

    public boolean search(String word) {
        TreeNode node = root;
        for (int i = 0; i < word.length(); i++) {
            int currChar = word.charAt(i) - 'a';
            if (node.next[currChar] == null) {
                return false;
            }
            node = node.next[currChar];
        }
        return node.isEnd;
    }

    public boolean startWith(String prefix) {
        TreeNode node = root;
        for (int i = 0; i < prefix.length(); i++) {
            int currChar = prefix.charAt(i) - 'a';
            if (node.next[currChar] == null) {
                return false;
            }
            node = node.next[currChar];
        }
        return true;
    }
}

File: Week_07/G20200343040105/Trie.cpp
//208. 实现 Trie (前缀树)

class Trie {
    Trie* next[26];
    bool isEnd;
public:
    /** Initialize your data structure here. */
    Trie() {
        for (int i = 0; i < 26; ++i) next[i] = NULL;
        isEnd = false;
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        Trie* node = this;
        for (char c : word) {
            if (node->next[c - 'a'] == NULL) {
                node->next[c - 'a'] = new Trie();
            }
            node = node->next[c - 'a'];
        }
        node->isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        Trie* node = this;
        for (char c : word) {
            if (node->next[c - 'a'] == NULL) return false;
            node = node->next[c - 'a'];
        }
        return node->isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        Trie* node = this;
        for (char c : prefix) {
            if (node->next[c - 'a'] == NULL) return false;
            node = node->next[c - 'a'];
        }
        return true;        
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */